torch_xla/csrc/lowering_context.h

```diff
@@ -110,6 +115,14 @@ class LoweringContext : public torch::lazy::LoweringContext {
       parameters_map_;
   std::vector<xla::XlaOp> root_tuple_;
   OutputMap<xla::XlaOp> emitted_outputs_;
+
+  // Stack frame index tables - we accumulate and write these to the HloModule
+  xla::StackFrameIndexProto indexes_;
+
```

Comment:
I am trying to think a way to not keep these states, and I think way to accomplish that is the following:

1. Here we keep one extra field `std::vector< std::vector< FrameInfo>*> finfos` here. 
2. In lowering_context.cpp; we just store the pointer `&nmeta.frame_info` here.
3. Have a function `void FillStackFrameIds( const std::vector< std::vector<FrameInfo>*>& finfos, StackFrameIndexProto* )` etc that, takes everything and computes Stack frame index; then; these maps can be local variable to that function. This function can be called in `BuildXLA`.
wdyt?

Reply:
One concern with the above is: if the torch::lazy::Node graph goes out of scope (is deleted) prior to lowering this will seg fault on frame info vector pointers.  Is the graph guaranteed to stay live until lowering is complete?

Reply:
Also realized this means we don't compute and lower the stack frame id into the HLO metadata at the moment of lowering a given node.  This means that we have to *also* somehow find that node after the fact as we are parsing the frames post node lowering.  I think this is a lot less efficient.  Let me take a look at the state storaga again ...

Reply:
So the storage here is actually pretty efficient: https://github.com/pytorch/xla/pull/5838/files#diff-04a457b5166774690501626cb60a8b2ae595a9abf4359f70f88a584ab750b274R122-R123 is only storing string_view objects (not strings) so this is just two pointers each. AFAIK (i.e. start and end).  So state is only really accumulating in the StackFrameIndex proto itself.  This means at the moment of lowering a node there are exactly two copies of the path and function strings.  If the nodes are quickly deleted after lowering this would limit the memory usage.

Reply:
I understand that we would like to move the code complexity out of this class.  I'd recommend that post shipping the change we make a note and start the process of building this into openxla: https://github.com/openxla/xla/blob/952d3cf39c3e3eeaa790cc1dd53423c8eb27d473/xla/client/xla_builder.h#L70

i.e. extend XlaBuilder / create a new helper that populates the underlying HloProto in the XlaComputation class.  This would be in line with how the rest of the HloProto is constructed and return the LoweringContext to a minimal form

However this will make no difference to memory overhead and needs to be a two-step to meet the 12/1 deadline for 2.1.2 and 2.2 releases.